### PR TITLE
MONGOID-4843 don't populate .changed_attributes setting has_and_belongs_to_many association

### PR DESCRIPTION
**Pull request by Neilshweky**
MONGOID-4843

### CODE REVIEW

File: lib/mongoid/fields.rb [hunk 1]
```diff
@@ -600,7 +600,7 @@ def create_field_getter(name, meth, field)
               write_attribute(name, field.eval_default(self))
             else
               value = field.demongoize(raw)
-              attribute_will_change!(name) if value.resizable?
+              attribute_will_change!(name) if value.resizable? && !field.foreign_key?
```
Comment (Neilshweky) [inline lib/mongoid/fields.rb:603]:
```suggestion
              attribute_will_change!(name) if value.resizable?
```

File: docs/reference/crud.txt [hunk 2]
```diff
@@ -863,6 +863,12 @@ the database up to the time it is saved. Any persistence operation clears the ch
    # Get the previous value for a field.
    person.name_was # "Alan Parsons"
 
+.. note::
+
+  Setting the associations on a document does not cause the ``changed`` or
+  ``changed_attributes`` hashes to be modified. This is true for all associations
+  whether referenced or embedded.
```
Comment (johnnyshields) [inline docs/reference/crud.txt:870]: FYI--In my app, we use a gem hack https://github.com/tablecheck/mongoid_relations_dirty_tracking that allows us to track this. For **embedded** (which are expected to behave like fields), there is a strong argument to add change tracking to Mongoid core in the future. (Not needed for referenced.)

File: spec/mongoid/changeable_spec.rb [hunk 3]
```diff
@@ -1012,12 +1012,29 @@
         person.user_accounts << user_account
       end
 
-      it 'returns a hash of changes' do
-        pending 'https://jira.mongodb.org/browse/MONGOID-4843'
+      it 'should not add to the changes or changed_attributes hash' do
+        person.changes.should == {}
+        person.changed_attributes.should == {}
+      end
```
Comment (johnnyshields) [inline spec/mongoid/changeable_spec.rb:1018]: Foreign key changes on `_ids` must continue to be tracked for both BelongsTo and HasAndBelongsToMany. The correct spec here is:

File: docs/reference/crud.txt (hunk 2 above)
Comment (Neilshweky) [reply docs/reference/crud.txt:870]: I agree, would you mind creating a ticket for this?

Comment (johnnyshields) [reply docs/reference/crud.txt:870]: I will do so, thanks!

File: spec/mongoid/changeable_spec.rb (hunk 3 above)
Comment (Neilshweky) [reply spec/mongoid/changeable_spec.rb:1018]: Hi @johnnyshields For referenced associations the `changes` hash was always the empty hash on change to the association, for every mongoid version that I looked at. The spec you provided would actually be a change to the current Mongoid functionality. Here's a code snippet if you'd like to verify that:
```
class User
  include Mongoid::Document

  has_and_belongs_to_many :addresses
end

class Address
  include Mongoid::Document

  has_and_belongs_to_many :users
end

a = Address.create!
u = User.create!

u.addresses << a
p u.changed
```

File: docs/reference/crud.txt (hunk 2 above)
Comment (johnnyshields) [reply docs/reference/crud.txt:870]: https://jira.mongodb.org/browse/MONGOID-5272

File: spec/mongoid/changeable_spec.rb (hunk 3 above)
Comment (johnnyshields) [reply spec/mongoid/changeable_spec.rb:1018]: @Neilshweky at least if I use `u.address_ids` for assignment, I see that it does appear in `u.changes`:

```ruby
u.address_ids << a._id
=> [BSON::ObjectId('6286663e071daa69cc155284')]

u.changes
=> {"address_ids"=>[nil, [BSON::ObjectId('6286663e071daa69cc155284')]]}
```

Comment (johnnyshields) [reply spec/mongoid/changeable_spec.rb:1018]: ah... ok. I keep forgetting `u.addresses` persists immediately, while `u.address_ids` does not. Hence `u.addresses` doesn't track the change (not even in `previous_changes`).

As long as  `u.address_ids << a._id` continues to track the change, we're fine.

We've discussed in other threads but the immediate persistence of relations when using `<<` is one thing we should really change. It causes so much confusion...

Comment (johnnyshields) [reply spec/mongoid/changeable_spec.rb:1018]: Thank you for taking time to make this script!

Comment (Neilshweky) [reply spec/mongoid/changeable_spec.rb:1018]: Ah yes, I was talking about the association itself while you were talking about the `_ids` field. Apologies for the confusion... The address_ids does still continue to be populated in the changes hash after this PR.

**Comment by p-mongo** [comment]
@Neilshweky Please retitle this PR to reflect the change being made.

File: spec/mongoid/changeable_spec.rb (hunk 3 above)
Comment (Neilshweky) [reply spec/mongoid/changeable_spec.rb:1018]: done!

**Review by p-mongo — APPROVED** [review]

**Pull request merged**